### PR TITLE
added flat hierarchy prompt / option for view templates

### PR DESCRIPTION
**Pull request by cyborgsimon**
app creation:
-  on app creation a prompt is shown asking if views should use a flat hierarchy or not
- default answer is no (to keep the previous behavior)
- if the answer is no: client/views/&lt;viewname&gt;/&lt;view name&gt;.{htmlcoffee} files are created
- if the answer is yes: client/views/&lt;viewname&gt;.{htmlcoffee} files are created

view creation:
- the value of the answer is stored in the config (flatHistoryView)
- on view creation the value (flatHistoryView) is read from the config
- if the answer is no: client/views/<name>/<name>.{htmlcoffee} files are created
- if the answer is yes: client/views/<name>.{htmlcoffee} files are created

unit tests:
- added unit tests for app creation
- added unit tests for view creation

### CODE REVIEW

**Comment by cmalven** [comment]
Thanks for the great PR @cyborgsimon! I especially appreciate that you created some tests. That's great!

The only change I'd like to make is to make is to make it especially clear in the project setup UI what the difference is between the two options. I think a great way to do this would be to store one of two options for the view hierarchy (`flat` or `nested`) instead of one boolean value, and you choose from one of these two values on project creation with `choices ` similarly to how you choose your CSS preprocessor. What do you think?

I'll leave some specific line comments on how I'd suggest we do this. Really appreciate your contribution!

File: app/index.js [hunk 1]
```diff
@@ -38,17 +38,27 @@ MeteorGenerator.prototype.askFor = function askFor() {
       message: 'Describe this project briefly'
     },
     {
+      type: 'confirm',
```
Comment (cmalven) [inline app/index.js]:
```
{
  type: 'list',
  name: 'viewHierarchy',
  message: 'How would you like your views to be organized?',
  choices: [
    { name: 'Nested: Each pair of .html and .coffee template files gets its own directory', value: 'nested' },
    { name: 'Flat: All .html and .coffee template files share the same directory', value: 'flat' }
  ],
  default: 'nested'
}
```

File: app/index.js [hunk 2]
```diff
@@ -38,17 +38,27 @@ MeteorGenerator.prototype.askFor = function askFor() {
...
       message: 'Choose a CSS Preprocessor',
       choices: ['stylus', 'less', 'bootstrap', 'none']
-    }
+    },
   ];
 
   this.prompt(prompts, function (props) {
     this.projectName = props.projectName;
     this.projectSummary = props.projectSummary;
+    this.flatHierarchyViews = props.flatHierarchyViews;
```
Comment (cmalven) [inline app/index.js]: this.viewHierarchy = props.viewHierarchy;

File: app/index.js [hunk 3]
```diff
@@ -38,17 +38,27 @@ MeteorGenerator.prototype.askFor = function askFor() {
...
     this.cssPreprocessor = props.cssPreprocessor;
+    
+    this.config.set('flatHierarchyViews', this.flatHierarchyViews);
```
Comment (cmalven) [inline app/index.js]: `this.config.set('viewHierarchy', this.viewHierarchy);`

File: app/index.js [hunk 4]
```diff
@@ -70,12 +80,19 @@ MeteorGenerator.prototype.client = function app() {
   this.mkdir('client');
   this.mkdir('client/vendor');
   this.mkdir('client/views');
-  this.mkdir('client/views/layout');
-  this.mkdir('client/views/loading');
+  if (this.flatHierarchyViews === false) {
```
Comment (cmalven) [inline app/index.js]: `if (this.viewHierarchy === 'nested') {`

File: app/index.js [hunk 5]
```diff
@@ -70,12 +80,19 @@ MeteorGenerator.prototype.client = function app() {
...
+    this.mkdir('client/views/layout');
+    this.mkdir('client/views/loading');
+  }
   this.copy('client/main.coffee', 'client/main.coffee');
   this.copy('client/router.coffee', 'client/router.coffee');
-  this.copy('client/views/layout/layout.html', 'client/views/layout/layout.html');
-  this.copy('client/views/loading/loading.html', 'client/views/loading/loading.html');
+  if (this.flatHierarchyViews === false) {
```
Comment (cmalven) [inline app/index.js]: `if (this.viewHierarchy === 'nested') {`

File: test/test-creation.js [hunk 6]
```diff
@@ -4,35 +4,79 @@
...
 
-    it('creates expected files', function (done) {
-        var expected = [
-            // add files you expect to exist here.
-            '.jshintrc',
-            '.editorconfig'
-        ];
+    it('creates app with view templates inside a directory', function (done) {
+      helpers.mockPrompt(this.app, {
+        'flatHierarchyViews': false
```
Comment (cmalven) [inline test/test-creation.js]: `'viewHierarchy': 'nested'`

File: test/test-creation.js [hunk 7]
```diff
@@ -4,35 +4,79 @@
...
+          'client/views/layout/layout.html',
+          'client/views/loading/loading.html',
+          'client/views/index/index.html',
+          'client/views/index/index.coffee',
+        ]);
+        done();
+      });
+    });
+
+    it('creates app with view templates with a flat hierary', function (done) {
```
Comment (cmalven) [inline test/test-creation.js]: Typo: "hierarchy"

File: test/test-creation.js [hunk 8]
```diff
@@ -4,35 +4,79 @@
...
-        helpers.mockPrompt(this.app, {
-            'someOption': true
-        });
-        this.app.options['skip-install'] = true;
-        this.app.run({}, function () {
-            helpers.assertFiles(expected);
-            done();
-        });
+      helpers.mockPrompt(this.app, {
+        'flatHierarchyViews': true
```
Comment (cmalven) [inline test/test-creation.js]: `'viewHierarchy': 'flat'`

File: test/test-creation.js [hunk 9]
```diff
@@ -4,35 +4,79 @@
...
+          'client/views/loading.html',
+          'client/views/index.html',
+          'client/views/index.coffee'
+        ]);
+        done();
+      });
+    });
+
+    it('creates view template inside a directory', function (done) {
+      this.view.config.set('flatHierarchyViews', false);
```
Comment (cmalven) [inline test/test-creation.js]: `this.view.config.set('viewHierarchy', 'nested');`

File: test/test-creation.js [hunk 10]
```diff
@@ -4,35 +4,79 @@
...
+        helpers.assertFile([
+          'client/views/testview/testview.html',
+          'client/views/testview/testview.coffee',
+        ]);
+        done();
+      });
+    });
+    
+    it('creates view template with a flat hierary', function (done) {
+      this.view.config.set('flatHierarchyViews', true);
```
Comment (cmalven) [inline test/test-creation.js]: Typo: 'hierarchy'

```
this.view.config.set('viewHierarchy', 'flat');
```

File: view/index.js [hunk 11]
```diff
@@ -6,14 +6,20 @@ var ViewGenerator = module.exports = function ViewGenerator(args, options, confi
   // By calling `NamedBase` here, we get the argument to the subgenerator call
   // as `this.name`.
   yeoman.generators.NamedBase.apply(this, arguments);
-
-  console.log('Creating a new ' + this.name + ' view.');
+  
+  console.log('Creating a new ' + this.name + ' view (flat: ' + this.flatHierarchyViews + ').');
```
Comment (cmalven) [inline view/index.js]: `console.log('Creating a new ' + this.name + ' view (' + this.viewHierarchy + ').');`

File: view/index.js [hunk 12]
```diff
@@ -6,14 +6,20 @@ var ViewGenerator = module.exports = function ViewGenerator(args, options, confi
...
 
 util.inherits(ViewGenerator, yeoman.generators.NamedBase);
 
 ViewGenerator.prototype.files = function files() {
-  this.mkdir('client/views/' + this.name);
-  this.template('_index.html', 'client/views/' + this.name + '/' + this.name + '.html');
-  this.template('_index.coffee', 'client/views/' + this.name + '/' + this.name + '.coffee');
+  var subpath = '';
+  
+  if (this.config.get('flatHierarchyViews') === false) {
```
Comment (cmalven) [inline view/index.js]: `if (this.config.get('viewHierarchy') === 'nested') {`

**Comment by cyborgsimon** [comment]
good idea. i made the requested changes.

**Comment by cmalven** [comment]
Merged and released as version 0.1.0 on npm. Thanks @cyborgsimon!

**Pull request merged**